### ISSUE TITLE
Capturing or passing a ref object to a worklet permanently breaks that ref object

### ISSUE THREAD

**Issue by chriscoomber** [software-mansion/react-native-reanimated#3670]
### Description

In short: capturing a ref object in a worklet causes it to be frozen, meaning that it can no longer be updated. This feels like incorrect behaviour - a frozen copy should be passed to the worklet, but the original should be left unfrozen. The current behaviour causes some very tricky and hard-to-debug issues; for example if I add a `console.log(myRef)` inside a worklet that wasn't capturing `myRef` before, I have actually changed (broken) the behaviour of my component, since `myRef` can no longer be updated. Adding a log shouldn't be a breaking change!

In long: In some code I was working on, I was capturing a ref (`MutableRefObject`, obtained from react's `useRef`) with a worklet (inside a handler passed into `useAnimatedGestureHandler`). I wanted the ref, rather than ref.current, because I wanted up-to-date contents of that ref at the moment the worklet is called - there might not have been a rerender since the ref value change; in other words I want to resolve `.current` in the worklet, not when the worklet was declared.

This gave me an unexpected result - after the worklet was first called, my ref became read-only _even in code running on the React Native thread_. Simply by capturing the ref in a worklet, I had frozen the ref for both threads, and could no longer update ref.current on either.

Here's a toy component so you can see what I mean:

```
import { useRef, useState } from 'react';
import { Button, View } from 'react-native';
import { runOnUI } from 'react-native-reanimated';

function MyComponent() {
  const [_, setFoo] = useState({});
  const forceRerender = () => setFoo({});

  const counterRef = useRef(0);
  console.log('(RN thread) render', counterRef.current, Object.isFrozen(counterRef));

  const worklet1 = () => {
    'worklet';
    const ref = counterRef;
    console.log('(UI thread) worklet1', ref.current);
  };

  const worklet2 = () => {
    'worklet';
    console.log('(UI thread) worklet2', counterRef.current);
  };

  return <View>
    <Button onPress={() => counterRef.current += 1} title={`Increment counter: ${counterRef.current}`}/>
    <Button onPress={forceRerender} title='Force rerender'/>
    <Button onPress={() => runOnUI(worklet1)()} title='Run worklet 1'/>
    <Button onPress={() => runOnUI(worklet2)()} title='Run worklet 2'/>
  </View>;
}
```

Here, we can increment a counter by clicking the top button, and force a rerender with the second button, so we can see the latest value. (I'm deliberately using useRef instead of useState for the sake of this example.)

`worklet1` captures the entire `counterRef`.

- Good thing: if you increment the counter and then click "Run worklet 1", you get the latest value of `counterRef`!
- Bad thing: after you click "Run worklet 1", the entire thing completely breaks. `counterRef` can no longer be incremented. This is the phenomenon I'm describing. worklet1 is not simply given a copy of the ref object at the moment it is called, the original ref object is frozen (`Object.isFrozen(counterRef) === true`).

`worklet2` captures just `counterRef.current`, evaluated at the time that `worklet2` is declared. This doesn't suffer from the bad thing above, but also doesn't benefit from the good thing, making it unsuitable for what I was trying to do.

Note that on web, there's no issue, since there's no sending of objects between thread..

### Steps to reproduce

This is a general issue - see the toy component above for a repro.

### Snack or a link to a repository

https://snack.expo.dev/rpPA4YXB7

### Reanimated version

2.9.1

### React Native version

0.69.5

### Platforms

Android, iOS

### Workflow

Expo managed workflow

### Build type

Debug mode

### Device

Real device

### Device model

Samsung Galaxy S20 FE

### Acknowledgements

Yes

**Comment by efstathiosntonas** [comment]
Maybe use `useAnimatedRef` instead of "plain" ref?

**Comment by tomekzaw** [comment]
Hey, if you need to share some value between JS and UI context, you need to use `useSharedValue` instead of `useRef`.

**Comment by chriscoomber** [comment]
@efstathiosntonas That won't work, that's not what `useAnimatedRef` is for. `useAnimatedRef` is only supported for taking refs of components, and here I'm using a number. Moreover, on the UI thread `useAnimatedRef` turns into a function which tries to return its component's viewTag (i.e. it no longer has a .current), which is great for calling `measure(viewTag)`, but not helpful for me here as I'm not using the ref to store a component. The snack link I posted is great for trying this stuff out (though I can't see logs made from the UI thread for some reason, you might need to wrap the console.log in a `runOnJS` to test it properly). I tried your suggestion and it crashed the app (as .current didn't exist) as expected.

@tomekzaw Yep, that makes sense. However...

You _are_ allowed to capture objects from the host context in a worklet, which allows one-way sending of data. From experimentation, it looks like this always performs a deep freeze on that object - freezing it for everyone. I don't think this is documented anywhere, and it's dangerous as it breaks the object for use back on the RN thread. You're right in the sense that you can always get around this by wrapping the object in a `SharedValue`, but still think this needs to at least heavily documented somewhere, to save the poor souls like me who spent a lot of time figuring out what went wrong. In principle, capturing objects from the host context in worklets is meant to be supported! Or even better it would be good if a frozen copy was sent to the worklet, leaving the original back on the RN thread unharmed.

I'm happy to call this a documentation bug, if it's not possible to send a frozen copy.

**Comment by tomekzaw** [comment]
> In short: capturing a ref object in a worklet causes it to be frozen, meaning that it can no longer be updated. This feels like incorrect behaviour - a frozen copy should be passed to the worklet, but the original should be left unfrozen.

I'm afraid this is intentional. Technically, we don't need to freeze objects passed to worklets (there's no requirement for doing so), but we do freeze them in order to avoid a case where you first send some value to worklet and then modify the original value in JS context. This would be difficult to debug indeed.

> Adding a log shouldn't be a breaking change!

Yes, but capturing a variable is.

> I wanted the ref, rather than ref.current, because I wanted up-to-date contents of that ref at the moment the worklet is called - there might not have been a rerender since the ref value change; in other words I want to resolve .current in the worklet, not when the worklet was declared.

This sounds like a perfect use-case for `useSharedValue`. Shared values are meant specifically for sharing some mutable state between JS and UI contexts, independently from React renders.

Under the hood, the current value of shared value is in fact kept in C++ structures (as `int`, `std::string` or `std::vector` etc.), with proper synchronization (because it can be written and read from two threads at once). When you assign to `.value` property, the original JS value is converted to C++ structures. When you access `.value` property, the JS value is instantiated in the proper runtime.

Note that due to JS VM limitations, it's not possible to "share" a single instance of an object between two runtimes. A single object is always bound to one runtime and cannot be accessed from other runtimes. Even if `useRef` would work as you initially expected, it would still not be possible to access or modify the same object (`ref`) from two different runtimes (JS and UI) due to technical limitations (or rather architectural decisions).

> You are allowed to capture objects from the host context in a worklet, which allows one-way sending of data.

Yes, to capture the current value. This is primarily used for some serializable constants, component state etc. Instead of "one-way sending of data", I would say "substitution". There is no "communication" here - instead, the worklet is recreated each time with a new value (the value is constant so it cannot be "modified" itself).

The only edge-case here I can think of here is capturing functions, e.g.:

```js
const [foo, setFoo] = useState('foo');

const worklet = () => {
  'worklet';
  runOnJS(setFoo)('bar');
};
```

However, under the hood, this is implemented using a shared value that holds a reference to the original function.

> From experimentation, it looks like this always performs a deep freeze on that object - freezing it for everyone. I don't think this is documented anywhere, and it's dangerous as it breaks the object for use back on the RN thread.

That's true. In most cases, if your code is truly functional, this is completely harmless. For mutable objects this will not work, that's why we freeze the object in order to warn early. Remember that you can always explicitly clone the object yourself and capture the copy (which will get frozen) so the original object is left intact.

**Comment by chriscoomber** [comment]
>>I wanted the ref, rather than ref.current, because I wanted up-to-date contents of that ref at the moment the worklet is called - there might not have been a rerender since the ref value change; in other words I want to resolve .current in the worklet, not when the worklet was declared.
>
>This sounds like a perfect use-case for useSharedValue. Shared values are meant specifically for sharing some mutable state between JS and UI contexts, independently from React renders.

Yep, I think this was me getting confused. Objects are captured when the function declared, not when it's called. Since the function is on a different thread, the objects are copied to the new thread, so my ref.current (from inside the function) won't change after the point where it's captured. I understand now that I should have used useSharedValue.

I'm still not sure about 2 things:

1. I still feel like it shouldn't freeze the object on the RN thread. I think the current behaviour is hard to debug, whereas if we just said "worklets are provided with deep copies of objects that are captured", I think that would make more sense. I get that you might have bugs where you were expected an object to update, but it didn't, but right now you have bugs where you captured an object for some innocent purpose, and then it loses its mutability. I think the latter is worse because it's unexpected, whereas sending copies of objects to the other thread is expected.
2. I definitely feel like this should be documented. I don't think animation code is often particularly functional, so I don't think it's often harmless.

Aside: My suggestion was to not freeze the object on the JS thread, and just send an immutable copy to the UI thread. Is it possible that the problem with that is more in "how to make a deep copy", vs "this would be hard to debug"? Usually, making a deep copy is up to the owner of the object. Rust has a concept of a Clone trait that needs to be implemented on the struct itself, for example. No library can copy your struct if it doesn't implement this. Rust also uses traits to specify whether something should be passed by copy instead of by move (Copy, requires Clone to be implemented) (JS doesn't have the concept of move), and whether it is safe to move to other threads (Send), or even have a reference be shared by different threads (Sync). When capturing an object in a closure (function), that closure implements Copy, Send or Sync if all of its captured objects do. Is it possible that you do a freeze on the object, because there's no way to reliably perform a deep copy (since there's no way for objects to have the trait Clone/Copy implemented on them)?

**Comment by louiszawadzki** [comment]
I had the same issue using refs inside a `Gesture.Tap().onStart` callback after adding `react-native-reanimated`.
Inside the callback I now do some stuff related to animations and some computations that use `ref` objects.

I had to read the thread in details to find how to solve my issue, I thought adding a small example could help the next readers, so here it is.

Here's my component before making it work:

```javascript
export const Component = () => {
  const [frequency, setFrequency] = useState(null);
  const lastTapDate = useRef(null);

  const size = useSharedValue(20);

  const animatedStyles = useAnimatedStyle(() => {
    return {
      height: size.value,
      width: size.value,
    };
  });

  const tap = Gesture.Tap().onStart(event => {
    size.value = 20;
    size.value = withSpring(300);

    // ❌ Here `lastTapDate.current` would always be undefined
    if (!lastTapDate.current) {
      lastTapDate.current = Date.now();
      return;
    }

    const pulseTime = (Date.now() - lastTapDate.current);
    setFrequency(Math.round(1 / pulseTime));
    lastTapDate.current = Date.now();
  });

  return (
    <>
      <View style={StyleSheet.absoluteFillObject}>
        {frequency !== null && <Text>{frequency}</Text>}
        <Animated.View style={[animatedStyles]} />
      </View>
      <GestureDetector gesture={tap}>
        <View style={{...StyleSheet.absoluteFillObject}} />
      </GestureDetector>
    </>
  );
};
```

And here's how I fixed it, by basically putting my refs into a `SharedValue` object. Note that passing the return from a `useRef` as the `SharedValue` won't work.

```javascript
export const Component = () => {
  const [frequency, setFrequency] = useState(null);
  // 1️⃣ Turn the refs into a `SharedValue`
  const refs = useSharedValue({ lastTapDate: null});

  const size = useSharedValue(20);

  const animatedStyles = useAnimatedStyle(() => {
    return {
      height: size.value,
      width: size.value,
    };
  });

  // 2️⃣ Define a function outside of the tap handler
  const updateFrequency = () => {
    // ✅ Now lastTapDate gets updated
    if (!refs.value.lastTapDate) {
      refs.value.lastTapDate = Date.now();
      return;
    }

    const pulseTime = (Date.now() - refs.value.lastTapDate);
    setFrequency(Math.round(1 / pulseTime));
    refs.value.lastTapDate = Date.now();
  }

  const tap = Gesture.Tap().onStart(event => {
    size.value = 20;
    size.value = withSpring(300);

    // 3️⃣ Call the new function inside `runOnJS`
    runOnJS(updateFrequency)();
  });

  return (
    <>
      <View style={StyleSheet.absoluteFillObject}>
        {frequency !== null && <Text>{frequency}</Text>}
        <Animated.View style={[animatedStyles]} />
      </View>
      <GestureDetector gesture={tap}>
        <View style={{...StyleSheet.absoluteFillObject}} />
      </GestureDetector>
    </>
  );
};
```

Hope this will help others :)

**Comment by mgcrea** [comment]
I have been fighting this exact same bug with for a few hours and did not find a solution so far. In my case it somehow breaks [react-navigation](https://reactnavigation.org/docs/getting-started/) internal state (frozen) even if the ref used is a const that has nothing to do with react-navigation!! 

I want/need to use a ref from a children component (exposed via `forwardRef` + `useImperativeHandle`). As soon as I read the ref in the parent component (like in a `useAnimatedReaction()` prepare call), it somehow freeze the react-navigation internal state! And breaks when I try to navigate away:

Here is the very simple repro code to save you a click:

```tsx
import { forwardRef, FunctionComponent, PropsWithChildren, useImperativeHandle, useRef } from "react";
import { SafeAreaView } from "react-native";
import { useAnimatedReaction } from "react-native-reanimated";

export const ReadOnlyRefScreen: FunctionComponent = () => {
  const ref = useRef<Handle>(null);

  useAnimatedReaction(
    () => ref.current,
    (next, prev) => {
      // console.log([prev, next]);
    }
  );

  return (
    <SafeAreaView style={{ flex: 1 }}>
      <ComponentWithRef ref={ref} />
    </SafeAreaView>
  );
};

type Handle = {
  foo: string;
};

const ComponentWithRef = forwardRef<Handle, PropsWithChildren>(function ComponentWithRef({ children }, ref) {
  useImperativeHandle(
    ref,
    () => {
      return {
        foo: "bar",
      };
    },
    []
  );
  return <>{children}</>;
});
```

Did set up a basic repro case that you can find here: [mgcrea/react-native-reanimated-sandbox](https://github.com/mgcrea/react-native-reanimated-sandbox/blob/master/src/screens/ReadOnlyRefScreen.tsx)

Any idea what could be causing this or how to find a workaround?

You would save my day as I'm loosing hair over this! Thanks!!

--- 

Video:

https://user-images.githubusercontent.com/108273/224832474-f198dff5-cef1-430e-8c61-8a5ce49ba246.mp4

--- 

Trace:

<img width="520" alt="Screenshot 2023-03-13 at 22 06 20" src="https://user-images.githubusercontent.com/108273/224832386-4e5b2032-3459-4544-a0ec-87f1586864c9.png">

**Comment by artemis-prime** [comment]
@tomekzaw  I see your reasoning, and its, well, reasonable!  But to @chriscoomber  's point, this should *definitely* be documented more thoroughly.  Had I not found this thread, I would have been banging my head against the wall wondering why a ref update that is meant to trigger a useEffect is never happening!

Another thought... Since you through an error in `src/reanimated2/sharables.ts` when you encounter an "`INNACESSIBLE_OBJECT`", why not do something similar for refs?  I only got on the right track because I tried to modify a mobx `observable` on the UI thread, which (very understandably) is not an accessible object.

Regardless, +100 for documenting this!

TL:DR; My ultimate solution was simple: 

```
  const onMenuAnimationDone = (opening: boolean) => {
    ui.setMenuOpen(opening) // mobx store
  }

    // called from onClick
  const openMenu = (opening: boolean): void => {
  
    menuAnimation.value = withTiming(
      opening ? 1 : 0, 
      {
        duration: 200,
        easing: opening ? Easing.in(Easing.exp) : Easing.out(Easing.exp),
      },
      () => {
        runOnJS(onMenuAnimationDone)(opening)
      }
    )
  }
```

**Comment by susonthapa** [comment]
I also ran into this issue. I used to store the height of `Animated.View` using the `onLayout` prop into a ref. It used to work with Reanimated 2. Today, I updated to Reanimated 3 and it broke the animation as the ref wasn't updating properly. I was able to figure out that the `onLayout` was causing the issue but didn't know why it worked with Reanimated 2. 

Thank god I found this thread. I was starting to question my understanding of ref in React 🥲.
Here is the [line](https://github.com/susonthapa/react-native-dribbble-sample/blob/85650a9d0c302c4d5d59579eb8add4b11340aa44/src/UploadButton/UploadButton.tsx#L196) of code that was causing the issue.

**Comment by tomekzaw** [comment]
@susonthapa Why are you using `onLayout` with shared values? It is asynchronous (requires a UI->JS jump) so there can be one or more frames of delay. The recommended way is to use [`measure`](https://docs.swmansion.com/react-native-reanimated/docs/advanced/measure/).

**Comment by susonthapa** [comment]
> @susonthapa Why are you using `onLayout` with shared values? It is asynchronous (requires a UI->JS jump) so there can be one or more frames of delay. The recommended way is to use [`measure`](https://docs.swmansion.com/react-native-reanimated/docs/advanced/measure/).

I don't remember why I decided to use shared values in `onLayout`, I guessed it just worked with Reanimated 2. In the code sample below, the `buttonWidth` is a ref and it doesn't update when using Reanimated 3. This used to work with Reanimated 2. I don't know if there is some behavior change with `onLayout` in Reanimated 3.

```
onLayout={(e) => {
                if (buttonWidth.current === 0) {
                    buttonWidth.current = e.nativeEvent.layout.width
                    progressTranslate.value = -buttonWidth.current + 24
                }
```

Thanks for the recommendation, I have learned quite a few things just going through the issue.
